modules/margin/src/main/java/com/opengamma/sdk/margin/v3/InvokerMarginClient.java

```diff
@@ -216,6 +221,49 @@ public MarginCalcResult calculate(Ccp ccp, MarginCalcRequest request) {
   }
 
   @Override
+  public MultiCcpMarginCalcResult calculate(
+      List<Ccp> ccps,
+      MarginCalcRequest request) throws ExecutionException, InterruptedException {
+
+    Map<Ccp, CompletableFuture<MarginCalcResult>> completableFutures = ccps.parallelStream()
+        .collect(Collectors.toMap(Function.identity(), ccp -> calculateAsync(ccp, request)));
+
+    List<MarginCalcResult> multiCcpMarginResults = new ArrayList<>();
+    HashMap<Ccp, MarginSummary> multiCcpMarginSummaries = new HashMap<>();
+    for (Map.Entry<Ccp, CompletableFuture<MarginCalcResult>> entry : completableFutures.entrySet()) {
```

Comment:
The general approach is to use the facilities of the executor to wait for all the futures to complete, perhaps with a timeout.  As it stands, the code will wait in the `get()` if the first in the map is slow, and later ones are faster. Here is a piece of common code to convert:

```
private static <T> CompletableFuture<List<T>> sequence(List<CompletableFuture<T>> futures) {
  CompletableFuture<Void> allDoneFuture =
      CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
  return allDoneFuture.thenApply(v ->
      futures.stream().
          map(future -> future.join()).
          collect(Collectors.<T>toList())
  );
}
```

although whether that handles exceptions in the way you want is open to question.